src/LinkedList.ts

```diff
@@ -2,54 +2,52 @@ import { fabric } from "fabric";
 import { Node } from "./Node";
 import { makeLine } from "./Utils";
 import Config = require("./Config");
+import { Variable } from "./Variable";
 
 export class LinkedList {
 
-    private head: Node = null;
+    private globalVars: { [key: string]: Variable };
+    private nodes: Node[];
     private canvas: fabric.Canvas;
 
-    constructor(canvasEl: HTMLCanvasElement, values:number[]=[]) {
+    constructor(canvasEl: HTMLCanvasElement, values: number[] = []) {
 
         this.canvas = new fabric.Canvas(canvasEl);
+        this.canvas.selection = false;
 
-        values = values.reverse();
+        // this allows us to draw circles using the coordinates of their centers
+        fabric.Object.prototype.originX = fabric.Object.prototype.originY = 'center';
+
+        this.nodes = [];
         for (let num of values) {
-            this.head = new Node(num, this.head);
+            this.nodes.push(new Node(num, this.canvas));
         }
 
-        this.draw()
-        
-        let nul = new fabric.IText('NULL', {
-            left: Config.LIST_X + (values.length * Config.NODE_SPACE),
-            top: Config.LIST_Y + (Config.NODE_SIZE / 2),
-            fill: '#black',
-        });
-        this.canvas.add(nul) 
+        for (let i = 0; i < this.nodes.length - 1; ++i) {
+            this.nodes[i].next.destination = this.nodes[i + 1];
+        }
+
+        this.globalVars = {
+            head: new Variable("head", this.canvas),
+        }
+        this.globalVars.head.value.destination = this.nodes[0];
+
+        this.draw();
 
-        let head = new fabric.IText('head', {
-            left: 20,
-            top: 20,
-            fill: '#black',
+        const self = this;
+        this.canvas.on('object:moving', function (e) {
+            self.draw();
+            self.canvas.renderAll();
         });
-        this.canvas.add(head) 
-
-        let line = makeLine([60, 60, 100, 140]);
-        let arrow1 = makeLine([95, 110, 100, 140]);
-        let arrow2 = makeLine([70, 120, 100, 140]);
-        this.canvas.add(line);
-        this.canvas.add(arrow1); 
-        this.canvas.add(arrow2);
-    
     }
 
     public draw() {
-        let leftEdge = Config.LIST_X;
-        let temp = this.head;
-        while (temp) {
-            temp.draw(this.canvas, leftEdge);
-            
-            leftEdge += Config.NODE_SPACE;
-            temp = temp.next;
+        for (let key in this.globalVars) {
+            this.globalVars[key].draw();
+        }
+
+        for (let node of this.nodes) {
+            node.draw();
```

Comment:
What if instead of doing this, we just tell the head to draw itself, but each time you draw a node, it also tells everything it points to to draw themselves??? This would make it REALLY easy when we want dangling parts of the diagram to disappear, because all we have to do is clear the canvas completely, then tell all the Global Pointers to draw themselves---anything not pointed to won't be redrawn. Does that seem like a good idea? You can tell me if its not.

Reply:
I don't think clearing the canvas will work with our current design. Currently, the Fabric objects that represent a node/pointer/variable are created in the constructor, and their positions are updated in the draw function (the draw function is called whenever anything changes, e.g. a node is dragged). If we clear the canvas, the draw function isn't going to recreate the Fabric objects.

Reply:
I'm not sure what you mean by "the draw function isn't going to recreate the Fabric objects." 
the Fabric objects should still all exist, then only the ones we want will be added drawn on the canvas.

Reply:
Another idea: maybe we have a list of *just* the nodes that don't have any pointers to them currently?